Add fleet-server secret file docs (#148)
elastic/ingest-docs: docs/en/ingest-management/fleet/fleet-server-secrets.asciidoc

[[fleet-server-secrets]]
= {fleet-server} Secrets

{fleet-server} configuration will contain secret values.
These values may be specified directly in the configuration or through secret files.
Command line args may be used to pass the values or file paths when running under the {agent}, or environment variables may be used if the {agent} is running in a container.

> 
> NOTE: Stand-alone {fleet-server} is under active development.
> 
> == Secret Values

Reviewer: ```suggestion
For examples of how to deploy secret files, refer to our <<secret-files-guide,Secret files guide>>.

NOTE: Stand-alone {fleet-server} is under active development.

[discrete]
== Secret Values
```
This `[discrete]` tag will keep the "Secret values" section on the same page as the top section. The "Secret values Guide" works nicely as a separate page, so we could add a link right here to make sure people find the examples.

If running {fleet-server} under the {agent} in a container (see <<elastic-agent-container>>), the environment variable `FLEET_SERVER_CERT_KEY_PASSPHRASE` may be used to specify the file path.

APM API Key::
The APM API Key may be used to gather APM data from {fleet-server}.

It can be directly specified in the instrumentation segment of the config:
[source,yaml]
----
inputs:
  - type: fleet-server
    instrumentation.api_key: my-apm-api-key
----

Or by a file with:
[source,yaml]
----
inputs:
  - type: fleet-server
    instrumentation.api_key_file: /path/to/apmAPIKey
----

It may also be specified by value with the environment variable `ELASTIC_APM_API_KEY`.

APM secret token::
The APM secret token may be used to gather APM data from {fleet-server}.

It can be directly specified in the instrumentation segment of the config:
[source,yaml]
----
inputs:
  - type: fleet-server
    instrumentation.secret_token: my-apm-secret-token
----

Or by a file with:
[source,yaml]
----
inputs:
  - type: fleet-server
    instrumentation.secret_token_file: /path/to/apmSecretToken
----

It may also be specified by value with the environment variable `ELASTIC_APM_SECRET_TOKEN`.

> == Secret files guide

Reviewer: ```suggestion
[[secret-files-guide]]
== Secret files guide
```
This would creates an ID so that we can link to here from the top of the "Fleet Server Secrets" page.

> The following guides provide step-by-step examples with best practices on how to deploy secret files directly on a host or through the kubernetes secrets engine.

Reviewer: ```suggestion
The following guide provides step-by-step examples with best practices on how to deploy secret files directly on a host or through the Kubernetes secrets engine.
```

[discrete]
== Secret Values

The following secret values may be used when configuring {fleet-server}.

Note that the configuration fragments shown below are specified either in the UI as part of the output specification or as part of the {fleet-server} integration settings.

`service_token`::
The `service_token` is used to communicate with {es}.
+
It may be specified in the configuration directly as:
+
[source,yaml]
----
output.elasticsearch.service_token: my-service-token
----
+
Or by a file with:
+
[source,yaml]
----
output.elasticsearch.service_token_path: /path/to/token-file
----
+
When you are running {fleet-server} under {agent} it can be specified with either the `--fleet-server-service-token` or the `--fleet-server-service-token-path` flag.
See <<elastic-agent-cmd-options,{agent} command reference>> for more details.
+
If you are <<elastic-agent-container,running {fleet-server} under {agent} in a container>>, the environment variables `FLEET_SERVER_SERVICE_TOKEN` or `FLEET_SERVER_SERVICE_TOKEN_PATH` may be used.

TLS private key::
The TLS private key is used to encrypt communications between {fleet-server} and {agent}.
See <<secure-connections>> for more details.
+

> While not reccomended, it may be specified in the configuration directly as:

Reviewer 2: ```suggestion
Although it is not recommended, you may specify the private key directly in the configuration as:
```

+
[source,yaml]
----
inputs:
  - type: fleet-server
    ssl.key: |
      ----BEGIN CERTIFICATE----
      ....
      ----END CERTIFICATE----
----
+

> Alternatively, the path to the private key can be provided with the same attribute:

Reviewer 2: ```suggestion
Alternatively, you can provide the path to the private key with the same attribute:
```

+
[source,yaml]
----
inputs:
  - type: fleet-server
    ssl.key: /path/to/cert.key
----
+

> When you are running {fleet-server} under {agent} the private key path may be provided with the `--fleet-server-cert-key` flag.

Reviewer 2: ```suggestion
When you are running {fleet-server} under {agent}, you can provide the private key path using with the `--fleet-server-cert-key` flag.
```

See <<elastic-agent-cmd-options,{agent} command reference>> for more details.
+

> If you are <<elastic-agent-container,running {fleet-server} under {agent} in a container>>, the environment variable `FLEET_SERVER_CERT_KEY` may be used to specify the private key path.

Reviewer 2: ```suggestion
If you are <<elastic-agent-container,running {fleet-server} under {agent} in a container>>, you can use the environment variable `FLEET_SERVER_CERT_KEY` to specify the private key path.
```

+
TLS private key passphrase::

> The private key passphrase is use to decrypt an encrypted private key file.

Reviewer 2: ```suggestion
The private key passphrase is used to decrypt an encrypted private key file.
```

+

> It can be specfied as a secret file in the configuration with:

Reviewer 2: ```suggestion
You can specify the passphrase as a secret file in the configuration with:
```

+
[source,yaml]
----
inputs:
  - type: fleet-server
    ssl.key_passphrase_path: /path/to/passphrase
----
+

> When you are running {fleet-server} under {agent} the passphrase path may be provided with the `--fleet-server-cert-key-passphrase-path` flag.

Reviewer 2: ```suggestion
When you are running {fleet-server} under {agent}, you can provide the passphrase path using the `--fleet-server-cert-key-passphrase-path` flag.
```

See <<elastic-agent-cmd-options,{agent} command reference>> for more details.
+

> If you are <<elastic-agent-container,running {fleet-server} under {agent} in a container>>, the environment variable `FLEET_SERVER_CERT_KEY_PASSPHRASE` may be used to specify the file path.

Reviewer 2: ```suggestion
If you are <<elastic-agent-container,running {fleet-server} under {agent} in a container>>, you can use the environment variable `FLEET_SERVER_CERT_KEY_PASSPHRASE` to specify the file path.
```

+
APM API Key::
The APM API Key may be used to gather APM data from {fleet-server}.
+

> It can be specified directly in the instrumentation segment of the configuration:

Reviewer 2: ```suggestion
You can specify it directly in the instrumentation segment of the configuration:
```

+
[source,yaml]
----
inputs:
  - type: fleet-server
    instrumentation.api_key: my-apm-api-key
----
+
Or by a file with:
+
[source,yaml]
----
inputs:
  - type: fleet-server
    instrumentation.api_key_file: /path/to/apmAPIKey
----
+

> It may also be specified by value with the environment variable `ELASTIC_APM_API_KEY`.

Reviewer 2: ```suggestion
You may specify the API key by value using the environment variable `ELASTIC_APM_API_KEY`.
```

APM secret token::
The APM secret token may be used to gather APM data from {fleet-server}.
+

> It can be directly specified in the instrumentation segment of the configuration:

Reviewer 2: ```suggestion
You can specify the secret token directly in the instrumentation segment of the configuration:
```

+
[source,yaml]
----
inputs:
  - type: fleet-server
    instrumentation.secret_token: my-apm-secret-token
----
+
Or by a file with:
+
[source,yaml]
----
inputs:
  - type: fleet-server
    instrumentation.secret_token_file: /path/to/apmSecretToken
----
+

> It may also be specified by value with the environment variable `ELASTIC_APM_SECRET_TOKEN`.

Reviewer 2: ```suggestion
You may also specify the token by value using the environment variable `ELASTIC_APM_SECRET_TOKEN`.
```

[[secret-files-guide]]
== Secret files guide

This guide provides step-by-step examples with best practices on how to deploy secret files directly on a host or through the Kubernetes secrets engine.

[[secret-filesystem]]
=== Secrets on filesystem

Secret files can be provisioned as plain text files directly on filesystems and referenced or passed through {agent}.

> We recommend the following steps in order to improve security.

Reviewer 2: ```suggestion
We recommend these steps to improve security.
```

> ==== File Permissions

Reviewer 2: ```suggestion
==== File permissions
```

File permissions should not allow for global read permissions.

> On MacOS and Linux, file ownership and file permissions can be set with the `chown` and `chmod` commands, respectively.

Reviewer 2: ```suggestion
On MacOS and Linux, you can set file ownership and file permissions with the `chown` and `chmod` commands, respectively.
```

> {fleet-server} runs as the `root` user on MacOS and Linux, so given a file named `mySecret`, it can be altered with:

Reviewer 2: ```suggestion
{fleet-server} runs as the `root` user on MacOS and Linux, so given a file named `mySecret`, you can alter it with:
```

[source,sh]
----
sudo chown root:root mySecret # set the user:group to root
sudo chmod 0600 mySecret      # set only the read/write permission flags for the user, clear group and global permissions.
----

On Windows, you can use `icacls` to alter the ACL list associated with the file:
[source,powershell]
----
Write-Output -NoNewline SECRET > mySecret          # Create the file mySecret with the contents SECRET
icacls .\mySecret /inheritance:d                   # Remove inherited permissions from file
icacls .\mySecret /remove:g BUILTIN\Administrators # Remove Administrators group permissions
icacls .\mySecret /remove:g $env:UserName          # Remove current user's permissions
----

==== Temporary filesystem

> A temporary filesystem (in RAM) can be used to hold secret files in order to improve security.

Reviewer 2: ```suggestion
You can use a temporary filesystem (in RAM) to hold secret files in order to improve security.
```

> These types of filesystems are normally not included in backups and cleared if the host is reset.

Reviewer 2: This statement seems ambiguous.  Do you mean that they are normally not included in backups and WILL BE cleared if the host is reset? 

For clarity, I'd recommend either: 
* These types of filesystems are normally not included in backups and ARE cleared if the host is reset, OR
* These types of filesystems are normally not included in backups and are NOT cleared if the host is reset.

My guess is the first option, but I don't want to guess incorrectly.

However, the filesystem and secret files need to be reprovisioned with every reset.

> On Linux `mount` can be used with the `tmpfs` filesystem to create a temporary filesystem in RAM:

Reviewer 2: ```suggestion
On Linux you can use `mount` with the `tmpfs` filesystem to create a temporary filesystem in RAM:
```

[source,sh]
----
mount -o size=1G -t tmpfs none /mnt/fleet-server-secrets
----

> On MacOS a combination of `diskutil` and `hdiutil` can be used to create a RAM disk:

Reviewer 2: ```suggestion
On MacOS you can use a combination of `diskutil` and `hdiutil` to create a RAM disk:
```

[source,sh]
----
diskutil erasevolume HFS+ 'RAM Disk' `hdiutil attach -nobrowse -nomount ram://2097152`
----

> For Windows systems, there are no built-in options to create a RAM disk.
> If this is required, a third party program is needed.

Reviewer 2: ```suggestion
Windows systems do not offer built-in options to create a RAM disk, but several third party programs are available.
```

==== Example

Here is a step by step guide for provisioning a service token on a Linux system:
[source,sh]
----
sudo mkdir -p /mnt/fleet-server-secrets
sudo mount -o size=1G -t tmpfs none /mnt/fleet-server-secrets
echo -n MY-SERVICE-TOKEN > /mnt/fleet-server-secrets/service-token
sudo chown root:root /mnt/fleet-server-secrets/service-token
sudo chmod 0600 /mnt/fleet-server-secrets/service-token
----

> NOTE: The `-n` flag is used with `echo` to prevent a newline character from being appended at the end of the secret. Care should be taken so that the secret file does not contain the trailing newline character.

Reviewer 2: ```suggestion
NOTE: The `-n` flag is used with `echo` to prevent a newline character from being appended at the end of the secret. Be sure that the secret file does not contain the trailing newline character.
```

> === Secrets in Containers

Reviewer 2: ```suggestion
=== Secrets in containers
```

> When using secret files directly in containers without using Kubernetes or another secrets management solution, the files can be passed into containers by mounting the file or directory.

Reviewer 2: ```suggestion
When you are using secret files directly in containers without using Kubernetes or another secrets management solution, you can pass the files into containers by mounting the file or directory.
```

> The file should be provisioned in the same manner as it is in <<secret-filesystem>> and mounted in read-only mode, for example when using Docker.

Reviewer 2: ```suggestion
Provision the file in the same manner as it is in <<secret-filesystem>> and mount it in read-only mode. For example, when using Docker.
```

> If using {agent} image:

Reviewer 2: ```suggestion
If you are using {agent} image:
```

For example, you can create the passphrase secret with:
[source,sh]
----
kubectl create secret generic fleet-server-key-passphrase \
  --from-literal=value=PASSPHRASE
----

And create the service token secret with:
[source,sh]
----
kubectl create secret generic fleet-server-service-token \
  --from-literal=value=SERVICE-TOKEN
----

Then include it in the pod specification, for example, when you are running {fleet-server} under {agent}:
[source,yaml]
----
spec:
  volumes:
  - name: key-passphrase
    secret:
      secretName: fleet-server-key-passphrase
  - name: service-token
    secret:
      secretName: fleet-server-service-token
  containers:
  - name: fleet-server
    image: docker.elastic.co/beats/elastic-agent
    volumeMounts:
    - name: key-passphrase
      mountPath: /var/secrets/passphrase
    - name: service-token
      mountPath: /var/secrets/service-token
    env:
    - name: FLEET_SERVER_CERT_KEY_PASSPHRASE
      value: /var/secrets/passphrase/value
    - name: FLEET_SERVER_SERVICE_TOKEN_PATH
      value: /var/secrets/service-token/value
----

==== {agent} Kubernetes secrets provider

> When you are running {fleet-server} under {agent} in Kuberenetes, {agent}'s <<kubernetes_secrets-provider>> may be used to insert a Kuberenetes secret directly into {fleet-server}'s configuration.

Reviewer 2: ```suggestion
When you are running {fleet-server} under {agent} in {k8s}, you can use {agent}'s <<kubernetes_secrets-provider>> to insert a {k8s} secret directly into {fleet-server}'s configuration.
```

Reviewer 2: Tip: You can use the `(k8s}` attribute for quicker entry and to prevent spelling errors. 
FYI: It resolves here: https://github.com/elastic/docs/blob/master/shared/attributes.asciidoc#L252
